### PR TITLE
[ONNX] 4 bits and 8 bits data-free WeightCompression

### PR DESCRIPTION
**Pull request by kshpv**
### Changes

Add a functional support of 4,8 bits weights compression for ONNX backend

There are images to demonstrate the look of same decompression subgraphs for ONNX vs OpenVINO:

This configuration demonstrates UINT8 weight quantization
image

This configuration demonstrates INT4 weight quantization with block size=64
image

The compression of the embeddings

image

Scale parameters of the embeddings quantization:

image

Zero point of the embeddings quantizations:

image

Accuracy results for a model from conformance tests - tinyllama/tinyllama-1.1b-step-50k-105b
The Similarity metrics between the compressed model vs FP32 baseline used

backend | INT8_SYM | INT8_ASYM
-- | -- | --
ONNX | 0.9316 | 0.95082
OpenVINO | 0.93018 | 0.95624
OpenVINO (scale   FP32) | 0.9316 | 0.95082

Performance results with [LLM_bench](https://github.com/openvinotoolkit/openvino.genai/tree/master/tools/llm_bench)
The 2nd token latency is presented via 3 measurements were done

backend | INT8_SYM | INT8_ASYM
-- | -- | --
ONNX | 44.96 ms/token | 44.96 ms/token
OpenVINO | 41.146 ms/token | 41.21 ms/token

### Reason for changes

Add new model backend

### Related tickets

163946

### Tests

Add unittetst for results of compression

### CODE REVIEW

File: nncf/onnx/graph/onnx_helper.py [hunk 1]
```diff
@@ -272,3 +288,26 @@ def is_node_has_shared_weight(
...
+    elif tensor.dtype == np.int8:
+        if np.max(tensor) > 7 or np.min(tensor) < -8:
+            msg = "Tensor values are not in [-8, 7]."
+            raise nncf.InternalError(msg)
+    else:
+        msg = f"Invalid weight dtype {tensor.dtype}."
+        raise nncf.InternalError(msg)
+    packed_tensor = np.ascontiguousarray(tensor)
+    packed_tensor = packed_tensor.reshape(-1, 2)
+    packed_tensor = packed_tensor[..., 1::2] << 4 | packed_tensor[..., ::2] & 15
```
Comment (alexsu52) [inline nncf/onnx/graph/onnx_helper.py:312]: PyTorch and ONNX has the same implementation of packing and unpacking 4bit in uint8. What do you think about standardizing this function and adding it to the tensor functions library?

File: nncf/quantization/algorithms/weight_compression/onnx_backend.py [hunk 2]
```diff
@@ -0,0 +1,327 @@
...
+        CompressWeightsMode.INT8_ASYM: onnx.TensorProto.UINT8,
+        CompressWeightsMode.INT4_SYM: onnx.TensorProto.INT4,
+        CompressWeightsMode.INT4_ASYM: onnx.TensorProto.UINT4,
+    }
+
+    def __init__(self, model: onnx.ModelProto):
+        super().__init__()
+        self.name_to_node_map = get_name_to_node_map(model)
+
+    def _get_weight_dtype(self, mode: CompressWeightsMode) -> onnx.TensorProto.DataType:
```
Comment (alexsu52) [inline nncf/quantization/algorithms/weight_compression/onnx_backend.py:66]: Please add a docstring for consistency.

File: nncf/quantization/algorithms/weight_compression/onnx_backend.py [hunk 3]
```diff
@@ -0,0 +1,327 @@
...
+        self, compressed_weight: Tensor, weight_shape: Tuple[int], dequantize_block_size: int
+    ) -> Tuple[np.ndarray, np.ndarray, Optional[np.ndarray]]:
+        """
+        Helper function to preprocess the tensor shapes for the compressed weight tensors to ONNX shapes expectations.
+        The function reshapes the weight tensor and squeezes the scale and zero point tensors based on the
+        dequantize_block_size parameter.
+        :param compressed_weight: The compressed weight tensor.
+        :param weight_shape: The original shape of the weight tensor.
+        :param dequantize_block_size: The block size for dequantization.
+        :return: A tuple containing the reshaped weight tensor, scale tensor, and zero point tensor (if applicable).
```
Comment (alexsu52) [inline nncf/quantization/algorithms/weight_compression/onnx_backend.py:83]:
```suggestion
        Helper function to preprocess the tensor shapes for the compressed weight tensors to ONNX shapes expectations.
        The function reshapes the weight tensor and squeezes the scale and zero point tensors based on the
        dequantize_block_size parameter.
        
        :param compressed_weight: The compressed weight tensor.
        :param weight_shape: The original shape of the weight tensor.
        :param dequantize_block_size: The block size for dequantization.
        :return: A tuple containing the reshaped weight tensor, scale tensor, and zero point tensor (if applicable).
```

**Review by alexsu52 — COMMENTED** [review]
Comments/questions from my side:
- Can you show what the ONNX decompression subgraph and the OpenIVNO decompression sub-graph look like for the same model with the same compression parameters?
- Are models supported that are presented as 2 files (graph and weights)?
- Please, collect performance and accuracy metrics and compare with OpenVINO backend.

File: nncf/onnx/graph/onnx_helper.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (kshpv) [reply nncf/onnx/graph/onnx_helper.py:312]: Do you propose to implement packing and unpacking methods for `nncf.Tensor`? If so, we've discussed with @alexsu52 and agreed not to do this because the places of usage packing and unpacking highly rely on backend-specific tensors such as `torch.tensor` or `np.array`.

[the remark this answers is not in the document]
Comment (anzr299) [reply nncf/onnx/graph/onnx_helper.py:312]: > Do you propose to implement packing and unpacking methods for nncf.Tensor?
Yes, I was talking about that.

> If so, we've discussed with @alexsu52 and agreed not to do this because the places of usage packing and unpacking highly rely on backend-specific tensors such as torch.tensor or np.array.

Oh alright👍

**Review by anzr299 — APPROVED** [review]

**Review by andrey-churkin — APPROVED** [review]
